### PR TITLE
siteconfig: Add siteconfig package for ClusterInstance custom resource

### PR DESCRIPTION
**Pull request by mpmaruthu**
Add siteconfig package for ClusterInstance custom resource.

Please help to review this PR when you have a time. Thanks !
/Peri M
to: @mcornea , @klaskosk , @trewest
cc: @yliu127

### CODE REVIEW

File: pkg/siteconfig/clusterinstance_test.go [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+		exists        bool
+		expectedError error
+	}{
+		{
+			exists:        true,
+			expectedError: nil,
+		},
+		{
+			exists:        false,
+			expectedError: fmt.Errorf("ClusterInstance cannot be deleted because it does not exist"),
```
Comment (klaskosk) [inline pkg/siteconfig/clusterinstance_test.go:174]: this needs to be updated since delete won't return an error now

File: .golangci.yml [hunk 2]
```diff
@@ -84,7 +84,7 @@ linters-settings:
       - ST1001
 
 linters:
-  enable:
+  disable:
```
Comment (klaskosk) [inline .golangci.yml:87]: also make sure this change isn't in the final PR

File: pkg/siteconfig/clusterinstance_test.go (hunk 1 above)
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance_test.go:174]: Good catch. Updated. Thanks!

File: .golangci.yml (hunk 2 above)
Comment (mpmaruthu) [reply .golangci.yml:87]: Okay, updated. Thanks!

File: go.mod [hunk 3]
```diff
@@ -210,3 +210,5 @@ replace (
 	github.com/portworx/sched-ops => github.com/portworx/sched-ops v1.20.4-rc1
 	k8s.io/client-go => k8s.io/client-go v0.29.4
 )
+
+replace github.com/openshift/assisted-service/models => github.com/openshift/assisted-service/models v0.0.0-20230831114549-1922eda29cf8
```
Comment (klaskosk) [inline go.mod:214]: could you remove this please and run `go mod tidy && go mod vendor` again`?

Comment (mpmaruthu) [reply go.mod:214]: Okay, done.

**Review by klaskosk — APPROVED** [review]
/lgtm

File: pkg/siteconfig/clusterinstance.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+				Name:      name,
+				Namespace: nsname,
+			},
+		},
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the ClusterInstance is empty")
+
+		return nil, fmt.Errorf("ClusterInstance 'name' cannot be empty")
```
Comment (trewest) [inline pkg/siteconfig/clusterinstance.go:56]:
```suggestion
		return nil, fmt.Errorf("clusterInstance 'name' cannot be empty")
```

File: pkg/siteconfig/clusterinstance.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the ClusterInstance is empty")
+
+		return nil, fmt.Errorf("ClusterInstance 'nsname' cannot be empty")
```
Comment (trewest) [inline pkg/siteconfig/clusterinstance.go:62]:
```suggestion
		return nil, fmt.Errorf("clusterInstance 'nsname' cannot be empty")
```

File: pkg/siteconfig/clusterinstance.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	if !builder.Exists() {
+		return nil, fmt.Errorf("ClusterInstance object %s does not exist in namespace %s", name, nsname)
```
Comment (trewest) [inline pkg/siteconfig/clusterinstance.go:66]:
```suggestion
		return nil, fmt.Errorf("clusterInstance object %s does not exist in namespace %s", name, nsname)
```

File: pkg/siteconfig/clusterinstance.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	if builder == nil {
+		glog.V(100).Infof("The %s builder is uninitialized", resourceCRD)
+
+		return false, fmt.Errorf("error: received nil %s builder", resourceCRD)
+	}
+
+	if builder.Definition == nil {
+		glog.V(100).Infof("The %s is undefined", resourceCRD)
+
+		builder.errorMsg = msg.UndefinedCrdObjectErrString(resourceCRD)
```
Comment (sebrandon1) [inline pkg/siteconfig/clusterinstance.go:174]:
```suggestion
		return false, fmt.Errorf(msg.UndefinedCrdObjectErrString(resourceCRD))
```

File: pkg/siteconfig/clusterinstance.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	if builder.apiClient == nil {
+		glog.V(100).Infof("The %s builder apiclient is nil", resourceCRD)
+
+		builder.errorMsg = fmt.Sprintf("%s builder cannot have nil apiClient", resourceCRD)
```
Comment (sebrandon1) [inline pkg/siteconfig/clusterinstance.go:180]:
```suggestion
		return false fmt.Errorf("%s builder cannot have nil apiClient", resourceCRD)
```

**Review by sebrandon1 — COMMENTED** [review]
Two minor comments

File: pkg/siteconfig/clusterinstance.go (hunk 4 above)
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:56]: Okay, done. Thanks!

File: pkg/siteconfig/clusterinstance.go (hunk 5 above)
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:62]: updated, thanks!

File: pkg/siteconfig/clusterinstance.go (hunk 6 above)
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:66]: updated, thanks!

File: pkg/siteconfig/clusterinstance.go (hunk 7 above)
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:174]: Updated, thanks!

File: pkg/siteconfig/clusterinstance.go (hunk 8 above)
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:180]: updated, thanks!

**Review by klaskosk — DISMISSED** [review]

**Review by sebrandon1 — APPROVED** [review]

File: pkg/siteconfig/clusterinstance.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	"github.com/openshift-kni/eco-goinfra/pkg/msg"
+	siteconfigv1alpha1 "github.com/openshift-kni/eco-goinfra/pkg/schemes/siteconfig/v1alpha1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ClusterInstanceBuilder provides struct for the ClusterInstance object.
+type ClusterInstanceBuilder struct {
+	Definition *siteconfigv1alpha1.ClusterInstance
```
Comment (kononovn) [inline pkg/siteconfig/clusterinstance.go:18]: Please add comments for public elements of the struct

**Review by kononovn — CHANGES_REQUESTED** [review]
1 comment to fix.

File: pkg/siteconfig/clusterinstance.go (hunk 9 above)
Comment (mpmaruthu) [reply pkg/siteconfig/clusterinstance.go:19]: @kononovn, I have updated the changes to address your review comments and submitted a new patch for another round of review. Thanks!

**Review by sebrandon1 — APPROVED** [review]

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**